### PR TITLE
CVSL-887 add auditing for licence condition service

### PR DESCRIPTION
**Pull request by gurps-singh**
This PR is to add audit events following changes to licence conditions. This introduces the logging of audit events to the `audit_event` table when conditions are added, removed or updated.

Previous PR  (https://github.com/ministryofjustice/create-and-vary-a-licence-api/pull/239) caused rebase error so created new PR.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt [hunk 1]
```diff
@@ -41,13 +46,37 @@ class LicenceConditionService(
     val entityStandardPssConditions = request.standardPssConditions.transformToEntityStandard(licenceEntity, "PSS")
 
     val username = SecurityContextHolder.getContext().authentication.name
+    val createdBy = communityOffenderManagerRepository.findByUsernameIgnoreCase(username)
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt:49]: wondering about what happens if this is a team member who hasn't done licence stuff before - can we guarantee that they will exist in the DB?

Comment (andrewrlee) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt:49]: This is probably fine - just requires a test

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt [hunk 2]
```diff
@@ -208,6 +349,70 @@ class LicenceConditionService(
...
+        "changes" to
+          removedConditions.map {
+            mapOf(
+              "conditionCode" to "",
+              "conditionType" to "",
+              "conditionText" to it,
+            )
+          },
+      )
+      auditEventRepository.saveAndFlush(
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt:391]: So for one call to update bespoke events you could end up with something like:

```yaml
  Event 1: 
       type: Removed bespoke conditions
       changes:
           typeOfChange/condition: delete bespoke conditions
           changes:
                - condition1 code, type text
                - condition2 code, type text
              
  Event 2: 
       type: Added bespoke conditions
       changes:
           typeOfChange/condition: added bespoke conditions
           changes:
                - condition3 code, type text
```

Feels like we could possible compress this structure structure a bit to make it more concise but contain the same info:
```yaml
  Event 1: 
       type: Updated bespoke conditions
       changes:
           changes:
                - condition1, ADDED, code, type, text
                - condition2, ADDED, code, type, text
                - condition3, REMOVED, code, type, text 
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt (hunk 1 above)
Comment (gurps-singh) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt:49]: Added an extra test

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt [hunk 3]
```diff
@@ -101,17 +151,45 @@ class LicenceConditionService(
...
 
     val username = SecurityContextHolder.getContext().authentication.name
+    val currentUser = communityOffenderManagerRepository.findByUsernameIgnoreCase(username)
+      ?: throw RuntimeException("Staff with username $username not found")
 
     // return all conditions except condition with submitted conditionId
     val revisedConditions = licenceEntity.additionalConditions.filter { it.id != conditionId }
 
+    val removedCondition =
+      licenceEntity.additionalConditions.filter { it.id == conditionId }.maxBy { it.id }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceConditionService.kt:161]: mmh.. could potentially use find here?

**Review by andrewrlee — APPROVED** [review]
LGTM

**Pull request merged**